### DISCUSSION TITLE
toNumber() Error

### DISCUSSION THREAD

**Discussion by a-znk** [wevm/wagmi#2867] [Q&A]
I'm in the process of upgrading an older dApp project, and I've run into the following error message: 

`Property 'toNumber' does not exist on type '{}'.ts(2339)`

Here is the code:

`const { data: maxMintAmountPerTxData } = useContractRead({
...contractConfig,
functionName: "maxMintAmountPerTx",
watch: true,
});`

`React.useEffect(() => {
    if (maxMintAmountPerTxData) {
      setmaxMintAmountPerTx(maxMintAmountPerTxData.toNumber());
    }
  }, [maxMintAmountPerTxData]);`

My understanding is that wagmi's transition to viem is the likely cause of this issue. I'm wondering what the best resolution would be. For whatever reason, this is really tripping me up. Thanks!

**Comment by jxom** [comment]
viem uses [native BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) instead of BigNumber.js under-the-hood, which do not have a toNumber method (they are just literal BigInt types).

**Reply by a-znk** [reply]
Thank you for the quick reply. I saw this in the docs. Do you have an alternative suggestion for this?

`setmaxMintAmountPerTx(maxMintAmountPerTxData.toNumber());`

I don't have much experience with BigInt... I worry I may be overthinking the fix. I've tried these, but it doesn't yield the result I need:

`setmaxMintAmountPerTx(BigInt(maxMintAmountPerTxData);`

Errors:
`Argument of type 'bigint' is not assignable to parameter of type 'SetStateAction<number>'.ts(2345)`
`Argument of type '{}' is not assignable to parameter of type 'string | number | bigint | boolean'.ts(2345)`

`setmaxMintAmountPerTx(BigInt(Number.maxMintAmountPerTxData);`

Errors: 
`Argument of type 'bigint' is not assignable to parameter of type 'SetStateAction<number>'.ts(2345)`
`Property 'maxMintAmountPerTxData' does not exist on type 'NumberConstructor'.ts(2339)`

An example would be appreciated, if possible. Thank you.

**Reply by jxom** [reply]
You will need to type your state as a `bigint` instead of a `number` ☺️

Example: 
```
const [value, setValue] = useState<bigint | undefined>()
```

**Reply by a-znk** [reply]
Thank you @jxom. I was able to get the error resolved doing the following: 

`const [maxMintAmountPerTx, setmaxMintAmountPerTx] = React.useState<bigint | undefined>();`

  `const { data: maxMintAmountPerTxData } = useContractRead({
    ...contractConfig,
    functionName: "maxMintAmountPerTx",
    watch: true,
  });`

  `React.useEffect(() => {
    if (maxMintAmountPerTxData) {
      setmaxMintAmountPerTx(BigInt(Number(maxMintAmountPerTxData)));
    }
  }, [maxMintAmountPerTxData]);`

Oddly though, when I try to display `maxMintAmountPerTxData` on the dApp, I receive this new error now: 

`Type 'unknown' is not assignable to type 'ReactNode'.ts(2322)`

`index.d.ts(1420, 9): The expected type comes from property 'children' which is declared here on type 'DetailedHTMLProps<HTMLAttributes<HTMLParagraphElement>, HTMLParagraphElement>'`

I haven't seen this before the viem/wagmi upgrade. Is it somehow related?
